src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,651 @@
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  std::optional<uint64_t> takeU64() {
+    if (auto t = peek()) {
+      if (auto n = t->getU64()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int64_t> takeS64() {
+    if (auto t = peek()) {
+      if (auto n = t->getS64()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int64_t> takeI64() {
+    if (auto t = peek()) {
+      if (auto n = t->getI64()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<uint32_t> takeU32() {
+    if (auto t = peek()) {
+      if (auto n = t->getU32()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int32_t> takeS32() {
+    if (auto t = peek()) {
+      if (auto n = t->getS32()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int32_t> takeI32() {
+    if (auto t = peek()) {
+      if (auto n = t->getI32()) {
+        ++lexer;
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<double> takeF64() {
+    if (auto t = peek()) {
+      if (auto d = t->getF64()) {
+        ++lexer;
+        return d;
+      }
+    }
+    return {};
+  }
+
+  std::optional<float> takeF32() {
+    if (auto t = peek()) {
+      if (auto f = t->getF32()) {
+        ++lexer;
+        return f;
+      }
+    }
+    return {};
+  }
+
+  std::optional<std::string_view> takeString() {
+    if (auto t = peek()) {
+      if (auto s = t->getString()) {
+        ++lexer;
+        return s;
+      }
+    }
+    return {};
+  }
+
+  std::optional<Name> takeName() {
+    // TODO: Move this to lexer and validate UTF.
+    if (auto str = takeString()) {
+      // Copy to a std::string to make sure we have a null terminator, otherwise
+      // the `Name` constructor won't work correctly.
+      // TODO: Update `Name` to use string_view instead of char* and/or to take
+      // rvalue strings to avoid this extra copy.
+      return Name(std::string(*str));
+    }
+    return {};
+  }
+
+  bool takeSExprStart(std::string_view expected) {
+    auto original = lexer;
+    if (takeLParen() && takeKeyword(expected)) {
+      return true;
+    }
+    lexer = original;
+    return false;
+  }
+
+  Index getPos() {
+    if (auto t = peek()) {
+      return lexer.getIndex() - t->span.size();
+    }
+    return lexer.getIndex();
+  }
+
+  [[nodiscard]] Err err(std::string reason) {
+    std::stringstream msg;
+    msg << lexer.position(lexer.getIndex()) << ": error: " << reason;
+    return Err{msg.str()};
+  }
+};
+
+// ===================
+// POD Utility Structs
+// ===================
+
+// The location and possible name of a module-level definition in the input.
+struct DefPos {
+  Name name;
+  Index pos;
+};
+
+struct GlobalType {
+  Mutability mutability;
+  Type type;
+};
+
+struct ImportNames {
+  Name mod;
+  Name nm;
+};
```

Comment:
It would be good to match the names here:

https://github.com/WebAssembly/binaryen/blob/0b66981d0e3aa958877b5c04769e75836497b4c0/src/wasm.h#L1666-L1668

Reply:
Hmm maybe. `mod`  and `nm` are the names used in the spec, and I think it might actually be more useful to match those in this file.

Reply:
Fair enough.